web/federate.go

```diff
@@ -179,7 +189,7 @@ func (h *Handler) federation(w http.ResponseWriter, req *http.Request) {
 			}
 			if l.Name == labels.MetricName {
 				nameSeen = true
-				if l.Value == lastMetricName {
+				if l.Value == lastMetricName && lastWasHistogram == isHistogram {
```

Comment:
This has the special challenge of what to do if there is a native histogram called `foo_metric` _and_ a conventional metric also called `foo_metric` (with different labels). This runs into the problem that the exposition format does not have the correct data model for the entirety of the metrics collected on a Prometheus server. So far, it didn't matter because we did not server the help string and the metric type through federation (everything is `untyped` and help-less).

Same problem when the same name has gauge and counter histograms.

I guess for now it's OK to only serve the metrics of the first seen metric type. But we need to clearly call that out.

Reply:
After some offline discussions, we decided that it is ok to have mixed type under a single metric family, since it is behind the native histogram feature flag and federation happens between two Prometheus instances.